include/nbl/asset/filters/CBlitImageFilter.h

```diff
@@ -588,22 +706,33 @@ class CBlitImageFilter : public CImageFilter<CBlitImageFilter<Swizzle,Dither,Nor
 		)
 		{
 			// TODO: investigate properly if its supposed be `size` or `size-1` (polyphase kinda shows need for `size`)
-			core::vectorSIMDi32 last(kernelX.getWindowSize().x,0,0,0);
+			core::vectorSIMDi32 last(kernelX.getWindowSize().x-1,0,0,0);
 			if (inImageType>=IImage::ET_2D)
-				last.y = kernelY.getWindowSize().x;
+				last.y = kernelY.getWindowSize().y-1;
 			if (inImageType>=IImage::ET_3D)
-				last.z = kernelZ.getWindowSize().x;
+				last.z = kernelZ.getWindowSize().z-1;
 			return last;
 		}
+		static inline core::vectorSIMDu32 getPhaseSupportLUTAxisOffsets(const core::vectorSIMDu32& phaseCount,
+			const CScaledImageFilterKernel<KernelX>& kernelX,
+			const CScaledImageFilterKernel<KernelY>& kernelY,
+			const CScaledImageFilterKernel<KernelZ>& kernelZ)
+		{
+			core::vectorSIMDu32 result;
+			result.x = 0u;
+			result.y = (phaseCount[0] * kernelX.getWindowSize().x) * sizeof(value_type);
+			result.z = ((phaseCount[0] * kernelX.getWindowSize().x) + (phaseCount[1] * kernelY.getWindowSize().y)) * sizeof(value_type);
+			return result;
+		}
 		// the blit filter will filter one axis at a time, hence necessitating "ping ponging" between two scratch buffers
-		static inline uint32_t getScratchOffset(const state_type* state, bool secondPong)
+		static inline uint32_t getScratchOffset(const state_type* state, bool secondPong,
+			const CScaledImageFilterKernel<KernelX>& scaledKernelX,
+			const CScaledImageFilterKernel<KernelY>& scaledKernelY,
+			const CScaledImageFilterKernel<KernelZ>& scaledKernelZ)
 		{
 			const auto inType = state->inImage->getCreationParameters().type;
-			const auto kernelX = state->contructScaledKernel(state->kernelX);
-			const auto kernelY = state->contructScaledKernel(state->kernelY);
-			const auto kernelZ = state->contructScaledKernel(state->kernelZ);
 
-			const auto window_end = getWindowEnd(state->inImage->getCreationParameters().type,kernelX,kernelY,kernelZ);
+			const auto window_end = getWindowEnd(state->inImage->getCreationParameters().type,scaledKernelX,scaledKernelY,scaledKernelZ);
 			// TODO: account for the size needed for coverage adjustment
```

Comment:
can you fix this TODO or is that fixed in #343 ?

Reply:
I don't get it. Coverage adjustment already works.

Reply:
so why is this TODO comment still here?

Reply:
it clearly about something....

Reply:
i think its because coverage adjustment uses some scratch too, but the amount it will use for the temp data is not taken into account in this `max`

Reply:
I don't see how it could be using any temporary memory.
Most of the operations are done in place. The places where it accesses intermediate/scratch storages are [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L337), [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L343) (though we might want to replace the 4 here with `MaxChannels`), and [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L356), and they all appear to be valid, in-bound indices.
Apart from this there are just two `atomic_int32_t` needed to keep the state for coverage adjustment.

I'm failing to understand the purpose of `alloc_decode_scratch` [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L422), but since it is common for both coverage adjustment and non-coverage adjustment cases, I doubt here is the issue.

Probably try asking the person who wrote this TODO?

Reply:
> (though we might want to replace the 4 here with MaxChannels),

please do.

> I'm failing to understand the purpose of alloc_decode_scratch [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L422), but since it is common for both coverage adjustment and non-coverage adjustment cases, I doubt here is the issue.

It's because the decode line is a captured-by-reference variable outside the lambdas which could run in parallel "for_each", now if we have multiple threads (or even SIMD lanes with `par_unseq`) hammering the same decode scratch space we'll get data races.

This is what the `(state->inExtent.width + real_window_size[0]) * std::thread::hardware_concurrency() * VectorizationBoundSTL` bound is for.

You can add that in a comment I guess.

> don't see how it could be using any temporary memory.
Most of the operations are done in place. The places where it accesses intermediate/scratch storages are [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L337), [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L343) (though we might want to replace the 4 here with MaxChannels), and [here](https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L356), and they all appear to be valid, in-bound indices.

Ok I went back into the code and figured out what its about.

Normally when you blit you'll decode your image into a 1-double-per-channel array of size:
`[output.W]x[inputH+kernelH_support]x[input.D]`
then as you move through the dimensions you get
`[output.W]x[output.H]x[input.D+kernelD_support]`
and then
`output.W x output.H x output.D`

The TODO is referring to the fact that coverage adjustment uses the scratch of the "next" axis to run `nth_element` 
https://github.com/Devsh-Graphics-Programming/Nabla/blob/a3c958b2659ddcfb90e3f6c22531a567f680251a/include/nbl/asset/filters/CBlitImageFilter.h#L337-356
its really an obnoxious use of memory.

The whole idea with `nth_element` was a dumb hack, it should be rewritten to mirror what you do on the GPU, which is a small histogram mapped to fixed precision floats (rescaled integers), prefix sum, binary search for "nth" that way.


So we should clean up my old sphagetti code to something like this:
```cpp
                enum E_SCRATCH_USAGE
                {
                	ESU_PHASE_SUPPORT_LUT=0,
                	ESU_DECODE_WRITE=1,
                	ESU_BLIT_X_AXIS_READ=ESU_DECODE_WRITE,
                	ESU_BLIT_X_AXIS_WRITE=2,
                	ESU_BLIT_Y_AXIS_READ=ESU_BLIT_X_AXIS_WRITE,
                	ESU_BLIT_Y_AXIS_WRITE=3,
                	ESU_BLIT_Z_AXIS_READ=ESU_BLIT_Y_AXIS_WRITE,
                	ESU_BLIT_Z_AXIS_WRITE=4,
                	ESU_ALPHA_VALUE_HISTOGRAM=5,
                	ESU_COUNT=6 // when you call `getScratchOffset(state,ESU_COUNT)` it would give the total scratch size needed
                };
		static inline uint32_t getScratchOffset(const state_type* state, const E_SCRATCH_USAGE usage)
		{
			const auto inType = state->inImage->getCreationParameters().type;
			const auto kernelX = state->contructScaledKernel(state->kernelX);
			const auto kernelY = state->contructScaledKernel(state->kernelY);
			const auto kernelZ = state->contructScaledKernel(state->kernelZ);
			
			// need to add the memory for phase support LUT
			const core::vectorSIMDu32 phaseCount = getPhaseCount(state->inExtentLayerCount, state->outExtentLayerCount,inType);
			const size_t phaseLUTElementCount = (phaseCount[0]*scaledKernelX.getWindowSize().x) + (phaseCount[1]*scaledKernelY.getWindowSize().y) + (phaseCount[2]*scaledKernelZ.getWindowSize().z);

			const auto real_window_size = getRealWindowSize(inType,kernelX,kernelY,kernelZ);
			core::vectorSIMDi32 intermediateExtent[3];
			getIntermediateExtents(intermediateExtent, state, real_window_size);
			assert(intermediateExtent[0].x == intermediateExtent[2].x);
			
			const size_t MaxParallelism = std::thread::hardware_concurrency() * VectorizationBoundSTL;
			
			// decode
			uint32_t pingBufferElementCount = (state->inExtent.width + real_window_size[0]) * MaxParallelism;
			// x-axis filter output
			uint32_t pongBufferElementCount = intermediateExtent[0].x*intermediateExtent[0].y*intermediateExtent[0].z;

			const uint32_t yWriteElementCount = intermediateExtent[1].x*intermediateExtent[1].y*intermediateExtent[1].z;
			if (inType>=2D && yWriteElementCount>pingBufferElementCount) // will there will be a y-pass
				pingBufferElementCount = yWriteElementCount;

			const uint32_t zWriteElementCount = intermediateExtent[2].x*intermediateExtent[2].y*intermediateExtent[2].z;
			if (inType>=3D && zWriteElementCount>pongBufferElementCount) // will there will be a z-pass
				pongBufferElementCount = zWriteElementCount;
			
			uint32_t retval = phaseLUTElementCount; // LUT is first and immutable
			switch (usage)
			{
				case ESU_PHASE_SUPPORT_LUT:
					retval = 0u; // LUT is first and immutable
					break;
				// ping
				case ESU_DECODE_WRITE:
				case ESU_BLIT_Y_AXIS_WRITE:
					// just don't add anything
					break;
				// pong (need to skip same memory used both for decode and Y-axis)
				case ESU_BLIT_X_AXIS_WRITE:
				case ESU_BLIT_Z_AXIS_WRITE:
					retval += pingBufferElementCount;
					break;
				case ESU_ALPHA_VALUE_HISTOGRAM:
					retval += pingBufferElementCount+pongBufferElementCount;
					[[fallthrough]];
				default: // ESU_COUNT
					// if we use coverage adjustment we needed some extra memory for histogram
					if (state->alphaSemantic==CState::EAS_REFERENCE_OR_COVERAGE)
						retval += kAlphaHistogramSize*MaxParallelism; // duplicate histogram, avoid atomics
					break;
			}
			return retval * sizeof(value_type) * MaxChannels;
		}

```
effectively putting the LUT first, DECODE/Y_WRITE next, then X_WRITE/Z_WRITE afterwards and the alpha value histogram last.

Then get rid of all other `getXXXOffset` functions and rephrase the `getXXXXSize` in terms of this one.

Reply:
Wondering why we don't use the `core::LinearAddressAllocator` here? 🤔

Reply:
This snippet has a bug because `ESU_COUNT` will directly go to the `default:` case and won't report the total scratch size correctly.
I'll restructure this a bit.

Reply:
I think its best to not rely on fallthroughs (except for the obvious ones --overlapping memory for different passes) and write it more explicitly.

Reply:
> This snippet has a bug because ESU_COUNT will directly go to the default: case and won't report the total scratch size correctly.
I'll restructure this a bit.

Another bug is, since the LUT data type can vary `{16, 32, 64}`, we would be overestimating the scratch size most of the times.

Reply:
I think [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/92228e74ddb54145e5865713b84780146a42f2d1/include/nbl/asset/filters/CBlitImageFilter.h#L51) can now go?
After implementing the new scratch usage regime, I don't see this being used at all.

Reply:
You have this in the above snippet:

```C++
if (state->alphaSemantic==CState::EAS_REFERENCE_OR_COVERAGE)
    retval += kAlphaHistogramSize*MaxParallelism; // duplicate histogram, avoid atomics
```

[I currently did it with atomics (`std::atomic_ref`)](https://github.com/Devsh-Graphics-Programming/Nabla/blob/01f0c5202fceb96d8603bbb563e150b0aace0c9d/include/nbl/asset/filters/CBlitImageFilter.h#L412), which we don't want.

I was wondering how do you want me to approach this, since you cannot control the number of threads in `std::execution` policy, and also (AFAIK) you can't get the number of threads invoked by it.
One way I could think of is:
1. Fork `std::thread::hardware_concurrency` number of `std::thread`s.
2. Make them operate on their own copy of histogram.
3. `std::thread::join`.
4. Make the main thread merge all the histogram into one, which will be used for further processing.

If we do decide to go down this route, then I don't understand why we have the multiplication with `MaxParallelism` above, because it also has a factor of `VectorizationBoundSTL` and AFAICT we're not using any SIMD anywhere (that could've been a possibility with execution policy, I guess)?

Reply:
> You have this in the above snippet:
> 
> ```c++
> if (state->alphaSemantic==CState::EAS_REFERENCE_OR_COVERAGE)
>     retval += kAlphaHistogramSize*MaxParallelism; // duplicate histogram, avoid atomics
> ```
> 
> [I currently did it with atomics (`std::atomic_ref`)](https://github.com/Devsh-Graphics-Programming/Nabla/blob/01f0c5202fceb96d8603bbb563e150b0aace0c9d/include/nbl/asset/filters/CBlitImageFilter.h#L412), which we don't want.
> 
> I was wondering how do you want me to approach this, since you cannot control the number of threads in `std::execution` policy, and also (AFAIK) you can't get the number of threads invoked by it. One way I could think of is:
> 
>     1. Fork `std::thread::hardware_concurrency` number of `std::thread`s.
> 
>     2. Make them operate on their own copy of histogram.
> 
>     3. `std::thread::join`.
> 
>     4. Make the main thread merge all the histogram into one, which will be used for further processing.
> 
> 
> If we do decide to go down this route, then I don't understand why we have the multiplication with `MaxParallelism` above, because it also has a factor of `VectorizationBoundSTL` and AFAICT we're not using any SIMD anywhere (that could've been a possibility with execution policy, I guess)?

It works for decode scratch memory, I've outlined this before
https://github.com/Devsh-Graphics-Programming/Nabla/blob/gpu_blit_filter/include/nbl/asset/filters/CBlitImageFilter.h#L506

We already "allocate" scratch per-thread & per-simd lane (yes the execution policy can make that possible, supposedly).

`MaxParallelism` is a good bound, and there's an [assert if it fails](https://github.com/Devsh-Graphics-Programming/Nabla/blob/gpu_blit_filter/include/nbl/asset/filters/CBlitImageFilter.h#L521) so we'll know if it does.

Reply:
> I think [this](https://github.com/Devsh-Graphics-Programming/Nabla/blob/92228e74ddb54145e5865713b84780146a42f2d1/include/nbl/asset/filters/CBlitImageFilter.h#L51) can now go? After implementing the new scratch usage regime, I don't see this being used at all.

yes it can.

Reply:
> > This snippet has a bug because ESU_COUNT will directly go to the default: case and won't report the total scratch size correctly.
> > I'll restructure this a bit.
> 
> Another bug is, since the LUT data type can vary `{16, 32, 64}`, we would be overestimating the scratch size most of the times.

Good catches, as for the second one, add a parameter for the LUT data type and make it compute accordingly.

Reply:
Resolved.